### DISCUSSION TITLE
Is there an example or doc of RA-TLS written in Rust?

### DISCUSSION THREAD

**Discussion by ChangoBuitrago** [gramineproject/gramine#1802] [Q&A]
I'm looking for an example or guidance on how to run a HTTPS API REST server written in Rust within the enclave, with the ability to perform RA-TLS on it. 
Thank you.

**Comment by dimakuv** [comment]
I guess the quick answer is: No, we are not aware of an RA-TLS implementation written in Rust.

Interestingly, the projects that I know of all seem to have C/C++ based implementations of SGX-attested TLS. See the list at https://github.com/CCC-Attestation/interoperable-ra-tls.

Why can't you use e.g. our Gramine RA-TLS libraries and write Rust wrappers around them with FFI?

**Reply by ChangoBuitrago** [reply]
Hi @dimakuv,
Thank you for the prompt reply, and I hope this finds you well.

I agree with you about FFI. Just FYI, I not only revised your `interoperable-ra-tls` list but also took the time to review Teaclave (which does have Rust SGX-SDK bindings), Enclaive.io, Mithril Security, Scone, and others.

Presently, popular interoperable languages include Rust, Go, TypeScript, and Python. While Gramine boasts the ability to support them, it lacks examples for RA-TLS in these languages, providing only a C/C++ example.

What concerns me the most is that even if I use FFI or rewrite the entire thing using my preferred language, it would be a waste of time because I don't have a supporting community behind me to review, contribute, and maintain it. Even then, I would have to go through all the years of know-how and hurdles that your team has already overcome.

Looking forward to your thoughts.
Warm regards,

**Comment by ChangoBuitrago** [comment]
@dimakuv. To make the discussion more tangible, consider the following Gramine Rust example. It's an HTTP rust server running in an enclave. Let's assume I want to RA the enclave before engaging with this server. 

https://github.com/gramineproject/gramine/blob/master/CI-Examples/rust/README.md

What are the suggested steps to do the RA using the current proposed C/C++ RA-TLS implementation?
Thank you!

**Reply by dimakuv** [reply]
> What are the suggested steps to do the RA using the current proposed C/C++ RA-TLS implementation?

Yep, fair enough. We indeed lack the bindings of RA-TLS to popular languages and frameworks.

@mkow @kailun-qin @woju Do you have any opinion on this?

**Reply by mkow** [reply]
I'd say it's too early for that and people want quite a lot of different languages, at least comparing to the number of engineers we have. We still change the API from time to time and making it backward-compatible across 5 different languages would be painful.

**Reply by kailun-qin** [reply]
> We indeed lack the bindings of RA-TLS to popular languages and frameworks.

Yes, I saw similar requests from several other users (incl. some production usage).

> We still change the API from time to time

Our API does not change that often?

> I'd say it's too early for that and people want quite a lot of different languages, at least comparing to the number of engineers we have.

Totally agree, but IMO maybe we could plan to start w/ one or two? Also, maybe this is a good topic to discuss in our community channels where we could gather more feedbacks on whether valuable and which to support?

**Reply by woju** [reply]
We have python bindings in Scaffolding already for client side. For server side we use `gramine-ratls-wrapper`, which is a separate process. Both are very much underdocumented.

**Reply by ChangoBuitrago** [reply]
Hi @dimakuv @mkow @kailun-qin @woju,

Thank you all for your comments. As a user and in my opinion, there are two key areas you could target.

The first is Web3 (given my background in this area). Here, I would strongly recommend providing support for either Rust or Go. The second area is AI. While Python remains the most popular language, Rust and Go are gaining traction as well.

An alternative approach would be to compile to WebAssembly (WASM) and offer examples for each language. This might be simpler than building bindings.

Finally, regardless of whether you choose WASM, bindings, or wrappers, a crucial missing piece is clear and concise documentation that explains the proper implementation of RA-TLS.

Thank you for considering my suggestions.

**Reply by kailun-qin** [reply]
> We have python bindings in Scaffolding already for client side. For server side we use gramine-ratls-wrapper, which is a separate process. Both are very much underdocumented.

Indeed, this reminds me that there may exist some third-party RA-TLS wrappers, e.g. https://github.com/konvera/gramine-ratls-golang.

**Reply by ChangoBuitrago** [reply]
@kailun-qin, Based on your shared wrapper link, I went through your repositories and found exactly what I was looking for. Once more, a couple of basic RA-TLS examples like the following one, plus some extra documentation, and you're good to go!

@dimakuv , How up-to-date is the following example regarding 'CCC-Attestation/interoperable-ra-tls'?

https://github.com/konvera/mutual-ratls-example/tree/main

**Reply by dimakuv** [reply]
> @dimakuv , How up-to-date is the following example regarding 'CCC-Attestation/interoperable-ra-tls'?

> https://github.com/konvera/mutual-ratls-example/tree/main

*Mutual attestation* has nothing to do with *interoperable attestation*:
- *Mutual* means that attestation happens in two directions: the client attests the server, and the server attests the client. Both the client and the server **must run the same runtime/framework/implementation**.
- *Interoperable* means that attestation can happen between client and server **of different runtimes/frameworks/implementations**.

So, *mutual* doesn't imply that client and server can be from different implementations/vendors.

Similarly, **interoperable** doesn't imply that client and server must attest each other (only the client attesting the server is enough).

**Reply by ChangoBuitrago** [reply]
Hi @dimakuv,
Thank you for your detailed explanation. Unfortunately, I expressed myself incorrectly. The issue is that I tested the Go `mutual-ratls-example` and found it to be outdated. I managed to address some basic issues such as missing trusted files in the manifest or symlinking the `libdcap_quoteprov.so`, etc. However, even after everything was in place, I still encountered the `Quote: verification failed with error SW_HARDENING_NEEDED failed`, regardless of the environment variables I defined as suggested in the Gramine official doc.

In summary, the Go wrapper and above Go example are not up to date with the Gramine team's library and the Gramine team's C/C++ RA-TLS example. This brings me back to our original discussion, where ultimately you may be right that, for now, the only alternative could be to either implement my own wrapper using FFI or find a solution directly over C/C++ libs.

**Reply by dimakuv** [reply]
> In summary, the Go wrapper and above Go example are not up to date with the Gramine team's library and the Gramine team's C/C++ RA-TLS example.

Yes, looks like that's the problem.

**Reply by ChangoBuitrago** [reply]
Thank you, @dimakuv, for your time and sharing your conclusion. Would you mind answering the following before closing the thread:

1. Can you or your team reference other project, even if undocumented, that may have already implemented or be implementing a Gramine Go wrapper?

2. Can a trusted machine (service, verifier) remotely verify an RA-TLS certificate provided by a client (attester) **without the necessary DCAP infrastructure installed and functioning correctly on the trusted machine**? If so, would it be feasible to develop a trusted machine (service, verifier) that can **run within a web browser** and still securely remote verify a DCAP attested client (attester)?

**Reply by dimakuv** [reply]
> Can you or your team reference other project, even if undocumented, that may have already implemented or be implementing a Gramine Go wrapper?

Actually, I wonder if @bigdata-memory or @m1ghtym0 knows of Go wrappers for Gramine RA-TLS.

From my side, I don't know any such projects.

> Can a trusted machine (service, verifier) remotely verify an RA-TLS certificate provided by a client (attester) without the necessary DCAP infrastructure installed and functioning correctly on the trusted machine? If so, would it be feasible to develop a trusted machine (service, verifier) that can run within a web browser and still securely remote verify a DCAP attested client (attester)?

Generally no. There must be some server which will ultimately say "Yeah, this RA-TLS certificate is fine" (more specifically, the SGX Quote embedded in the cert is fine).

There are projects like Edgeless Marblerun, MS Azure Attestation (MAA) and Intel Trust Authority (ITA), that make this task simpler. But it still requires the infrastructure on your local machine/cluster.

**Reply by ChangoBuitrago** [reply]
Thank you, @dimakuv. It's all very clear now. Do you have any insights into whether there will be a replacement for the EPID attestation flow?

**Reply by kailun-qin** [reply]
> Do you have any insights into whether there will be a replacement for the EPID attestation flow?

Sorry that I don't quite get this question, but do you mean https://www.intel.com/content/www/us/en/developer/articles/technical/software-security-guidance/resources/sgx-ias-using-epid-eol-timeline.html?

**Reply by ChangoBuitrago** [reply]
Yes, @kailun-qin, thank you. Based on your link and disregarding DCAP, the Intel Trust Authority seems like the EPID replacement. Is this something Gramine supports or will support?

**Reply by kailun-qin** [reply]
> Is this something Gramine supports or will support?

@ChangoBuitrago: Yes, there is an ongoing effort. Pls see https://github.com/gramineproject/contrib/issues/73 and https://github.com/gramineproject/contrib/pull/75 for details.

**Reply by ChangoBuitrago** [reply]
Thank you for your guidance @kailun-qin. I will definitely delve into it.

**Reply by dinvlad** [reply]
Sorry to revive this conversation – I'm facing a similar challenge, where I'd like to use a client running entirely in the browser (and compiled down to WASM), and a server running inside a secure enclave. Since it's been over a year now, is it now possible for such a client to quote-verify the server in such scenario? The core challenge seems to be compiling relevant pieces from Gramine and its dependencies down to WASM.

Perhaps @ChangoBuitrago you found a solution since then? Thank you 🙏
